src/open_clip/flava_model.py

```diff
@@ -127,164 +294,258 @@ def __init__(
 
         # Text encoder
+        text_cfg = FLAVATextCfg(**text_cfg) if isinstance(text_cfg, dict) else text_cfg
         self.text = _build_text_tower(embed_dim, text_cfg, quick_gelu, cast_dtype)
-        # self.text_masked_head = nn.Linear(embed_dim, self.text.config.vocab_size)
+        text_context_length = text_cfg.context_length  # includes CLS_T token
 
-        # Multimodal encoder
-        act_layer = QuickGELU if quick_gelu else nn.GELU
-        norm_layer = LayerNormFp32 if cast_dtype in (torch.float16, torch.bfloat16) else LayerNorm
-        self.multimodal = EmbeddingTransformer(
-            context_length=multimodal_cfg.context_length,
-            width=multimodal_cfg.width,
-            heads=multimodal_cfg.heads,
-            layers=multimodal_cfg.layers,
-            mlp_ratio=multimodal_cfg.mlp_ratio,
-            ls_init_value=multimodal_cfg.ls_init_value,
-            output_dim=embed_dim,
-            act_layer=act_layer,
-            norm_layer=norm_layer,
+        # unimodal MLM
+        self.mlm_head = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 4),
+            nn.ReLU(),
+            nn.Linear(embed_dim * 4, self.text.config.vocab_size),
         )
-        self.mm_masked_head = nn.Linear(embed_dim, self.text.config.vocab_size)
-        self.itm_head = nn.Linear(embed_dim, 1)
 
+        # unimodal MAE
+        self.mae_mask_ratio = vision_cfg.mae_mask_ratio
+        self.patch_mask_token = nn.Parameter(vision_cfg.mae_decoder_width ** -0.5 * torch.randn(vision_cfg.mae_decoder_width))
+        self.mae_decoder = _build_vision_mae_decoder(vision_cfg, quick_gelu, cast_dtype)
+
+        # Multimodal encoder
+        multimodal_context_length = visual_context_length + text_context_length + 1  # +1 for CLS_M token
+        multimodal_cfg = FLAVAMultimodalCfg(**multimodal_cfg) if isinstance(multimodal_cfg, dict) else multimodal_cfg
+        self.multimodal = _build_multimodal_tower(
+            embed_dim,
+            multimodal_cfg,
+            multimodal_context_length,
+            quick_gelu,
+            cast_dtype,
+        )
         self.image_to_mm_projection = nn.Linear(embed_dim, multimodal_cfg.width)
         self.text_to_mm_projection = nn.Linear(embed_dim, multimodal_cfg.width)
 
+        # Cross-modal MLM
+        self.mm_mlm_head = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 4),
+            nn.GELU(),
+            nn.Linear(embed_dim * 4, self.text.config.vocab_size),
+        )
+
+        # Cross-modal MAE
+        self.mm_patch_mask_token = nn.Parameter(multimodal_cfg.width ** -0.5 * torch.randn(multimodal_cfg.width))
+        self.mm_mae_head = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 4),
+            nn.GELU(),
+            nn.Linear(embed_dim * 4, vision_cfg.patch_size ** 2 * 3, bias=True),  # patch reconstruction
+        )
+
+        # ITM
+        self.itm_head = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 4),
```

Comment:
Where did this `* 4` number came from? Are you taking inspiration from transformer's MLP block?

Reply:
yes